torch/package/package_exporter.py

```diff
@@ -246,78 +338,101 @@ def _write_dep_graph(self, failing_module=None):
         arg = quote(template, safe="")
         return f"https://dreampuf.github.io/GraphvizOnline/#{arg}"
 
-    def _get_source_of_module(self, module: types.ModuleType) -> str:
+    def _get_source_of_module(self, module: types.ModuleType) -> Optional[str]:
         filename = getattr(module, "__file__", None)
         result = (
             None
             if filename is None or not filename.endswith(".py")
             else linecache.getlines(filename, module.__dict__)
         )
+
         if result is None:
-            extra = ""
-            if self.verbose:
-                extra = f" See the dependency graph for more info: \n{self._write_dep_graph(module.__name__)}"
-            raise ValueError(
-                f'cannot save source for module "{module.__name__}" because '
-                f'its source file "{filename}" could not be found.{extra}'
-            )
+            return None
+
         return "".join(result)
 
     def require_module_if_not_provided(self, module_name: str, dependencies=True):
-        if self._module_is_already_provided(module_name):
+        if (
+            module_name in self.dependency_graph
+            and self.dependency_graph.nodes[module_name].get("provided") is True
+        ):
             return
-        self.require_module(module_name, dependencies)
-
-    def require_module(self, module_name: str, dependencies=True):
-        """This is called by dependencies resolution when it finds that something in the package
-        depends on the module and it is not already present. It then decides how to provide that module.
-        The default resolution rules will mark the module as extern if it is part of the standard library,
-        and call :meth:`save_module` otherwise. Clients can subclass this object
-        and override this method to provide other behavior, such as automatically mocking out a whole class
-        of modules"""
-
-        root_name = module_name.split(".", maxsplit=1)[0]
-        if self._can_implicitly_extern(root_name):
+
+        if self._can_implicitly_extern(module_name):
+            # TODO we are externing individual modules instead just the top level package
             if self.verbose:
                 print(
-                    f"implicitly adding {root_name} to external modules "
+                    f"implicitly adding {module_name} to external modules "
                     f"since it is part of the standard library and is a dependency."
                 )
-            self._save_extern_module(root_name)
+            self.dependency_graph.add_node(
+                module_name, action=_ModuleProviderAction.EXTERN, provided=True
+            )
             return
 
-        for i, (pattern, action, _) in enumerate(self.patterns):
+        for pattern, action, _ in self.patterns:
             if pattern.matches(module_name):
```

Comment:
Is it possible for a module name to match multiple patterns? I think that would be user error, but this code will take the first rule that matches and ignore the others. Should we detect this case and provide a better error message?

E.g what if some module matches a generic intern glob as well as a more specific deny glob, but the intern happened to have been "declared" first?

Reply:
It is intended behavior (and what currently happens) for patterns to match in order, with the first one winning. This allows people to specify catch-alls near the bottom. Our docs should make this clear though.